torch/csrc/jit/ir.h

```diff
@@ -757,37 +744,65 @@ friend struct Node;
     delete *it;
     all_nodes.erase(it);
   }
+  void freeValue(Value * v) {
+    auto it = all_values.find(v);
+    JIT_ASSERT(it != all_values.end());
+    all_values.erase(it);
+  }
 };
 
+inline Value::Value(Node * node_, size_t offset_)
+: node_(node_),
+  offset_(offset_),
+  unique_(node_->graph_->next_unique_++),
+  stage_(node_->graph_->new_node_stage_) {
+  node_->graph_->all_values.emplace(this);
+}
+
+inline Graph * Value::owningGraph() {
+  return node()->owningGraph();
+}
+
+inline const Graph * Value::owningGraph() const {
+  return node()->owningGraph();
+}
+
+inline void Value::replaceAllUsesWith(Value * newValue) {
+  JIT_ASSERT(owningGraph() == newValue->owningGraph());
+  for(auto u : uses()) {
+    u.user->inputs_[u.offset] = newValue;
+    newValue->uses_.push_back(u);
+  }
+  uses_.clear();
+}
+
 inline Node::Node(Graph * graph_, NodeKind kind_) :
   kind_(kind_),
   graph_(graph_),
-  unique_(graph_->next_unique_++),
-  stage_(graph_->new_node_stage_),
-  type_(getInitialType(kind_)) {
+  stage_(graph_->new_node_stage_) {
   graph_->all_nodes.emplace(this);
 }
 
+inline void Node::eraseOutput(size_t i) {
+  JIT_ASSERT(i < outputs_.size());
+  JIT_ASSERT(outputs_[i]->uses().size() == 0);
+  Value * n = outputs_[i];
+  outputs_.erase(outputs_.begin() + i);
+  owningGraph()->freeValue(n);
+  for(size_t j = i; j < outputs_.size(); j++) {
+    outputs_[j]->offset_--;
+  }
+}
+
 inline void Node::destroy() {
```

Comment:
Since destroy now also gets rid of the outputs, we should add to its comment that it invalidates any Value pointers to the output in question. Most importantly, when you see the idiom `value->node()->destroy()`, `value` must not be used afterwards. (Not for this PR, but if people keep using invalidated value nodes, we should seriously consider keeping them around but marking them dead, so that we can give assert failures when someone tries to use them.)

Reply:
A good point, I'll add a comment.